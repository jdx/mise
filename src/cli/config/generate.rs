use std::path::PathBuf;

use crate::config::{Config, Settings};
use crate::file;
use crate::file::display_path;
use clap::ValueHint;
use miette::Result;

/// [experimental] Generate an .mise.toml file
#[derive(Debug, clap::Args)]
#[clap(visible_alias = "g", verbatim_doc_comment, after_long_help = AFTER_LONG_HELP)]
pub struct ConfigGenerate {
    /// Output to file instead of stdout
    #[clap(long, short, verbatim_doc_comment, value_hint = ValueHint::FilePath)]
    output: Option<PathBuf>,
}

impl ConfigGenerate {
    pub fn run(self) -> Result<()> {
        let _ = Config::try_get()?;
        let settings = Settings::try_get()?;
        settings.ensure_experimental()?;
        let doc = r#"
# # mise config files are hierarchical. mise will find all of the config files
# # in all parent directories and merge them together.
# # You might have a structure like:
#
# * ~/work/project/.mise.toml  # a config file for a specific work project
# * ~/work/.mise.toml          # a config file for projects related to work
# * ~/.config/mise/config.toml # the global config file
# * /etc/mise/config.toml      # the system config file
#
# # This setup allows you to define default versions and configuration across
# # all projects but override them for specific projects.
#
# # set arbitrary env vars to be used whenever in this project or subprojects
# [env]
# NODE_ENV = "development"
# NPM_CONFIG_PREFIX = "~/.npm-global"
# EDITOR = "code --wait"
#
# mise.file = ".env"                # load vars from a dotenv file
# mise.path = "./node_modules/.bin" # add a directory to PATH
#
# [tools]
# terraform = '1.0.0'       # specify a single version
# erlang = '26'             # specify a major version only
# node = 'ref:master'       # build from a git ref
# node = 'path:~/.nodes/14' # BYO â€“ specify a non-mise managed installation
#
# # newest with this prefix (typically exact matches don't use the prefix)
# go = 'prefix:1.16'
#
# # multiple versions will all go into PATH in the order specified
# # this is helpful for making `python311` and `python310` available
# # even when `python` and `python3` point to a different version
# python = ['3.12', '3.11', '3.10']
#
# # some plugins can take options like python's virtualenv activation
# # with these, mise will automatically setup and activate vevs when entering
# # the project directory
# python = {version='3.12', virtualenv='.venv'}
# poetry = {version='1.7.1', pyproject='pyproject.toml'}
#
# [plugins]
# # specify a custom repo url so you can install with `mise plugin add <name>`
# # note this will only be used if the plugin is not already installed
# python = 'https://github.com/asdf-community/asdf-python'
#
# [alias.node]
# # setup a custom alias so you can run `mise use -g node@work` for node-16.x
# work = '16'
"#
        .trim();
        if let Some(output) = &self.output {
            info!("writing to {}", display_path(output));
            file::write(output, doc)?;
        } else {
            miseprintln!("{doc}");
        }

        Ok(())
    }
}

// TODO: fill this out
static AFTER_LONG_HELP: &str = color_print::cstr!(
    r#"<bold><underline>Examples:</underline></bold>
  $ <bold>mise cf generate > .mise.toml</bold>
  $ <bold>mise cf generate --output=.mise.toml</bold>
"#
);

#[cfg(test)]
mod tests {
    use std::env;

    #[test]
    fn test_generate() {
        with_settings!({
            let out = assert_cli!("config", "generate");
            for line in out.lines() {
                assert!(line.len() < 80);
            }
            assert_cli_snapshot!("cfg", "generate");
        });
    }
}
