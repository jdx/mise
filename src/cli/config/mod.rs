use clap::Subcommand;
use eyre::Result;

mod generate;
mod ls;

/// [experimental] Manage config files
#[derive(Debug, clap::Args)]
#[clap(visible_alias = "cfg")]
pub struct Config {
    #[clap(subcommand)]
    command: Option<Commands>,

    /// Do not print table header
    #[clap(long, alias = "no-headers", verbatim_doc_comment)]
    no_header: bool,
}

#[derive(Debug, Subcommand)]
enum Commands {
    Ls(ls::ConfigLs),
    Generate(generate::ConfigGenerate),
}

impl Commands {
    pub fn run(self) -> Result<()> {
        match self {
            Self::Ls(cmd) => cmd.run(),
            Self::Generate(cmd) => cmd.run(),
        }
    }
}

impl Config {
    pub fn run(self) -> Result<()> {
        let cmd = self.command.unwrap_or(Commands::Ls(ls::ConfigLs {
            no_header: self.no_header,
        }));

        cmd.run()
    }
}
