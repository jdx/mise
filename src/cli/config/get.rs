use crate::cli::config::top_toml_config;
use crate::file::display_path;
use eyre::bail;
use std::path::PathBuf;

/// Display the value of a setting in a mise.toml file
#[derive(Debug, clap::Args)]
#[clap(after_long_help = AFTER_LONG_HELP, verbatim_doc_comment)]
pub struct ConfigGet {
    /// The path of the config to display
    pub key: Option<String>,

    /// The path to the mise.toml file to edit
    ///
    /// If not provided, the nearest mise.toml file will be used
    #[clap(short, long)]
    pub file: Option<PathBuf>,
}

impl ConfigGet {
    pub fn run(self) -> eyre::Result<()> {
        let mut file = self.file;
        if file.is_none() {
            file = top_toml_config();
        }
        if let Some(file) = file {
            let config: toml::Value = std::fs::read_to_string(&file)?.parse()?;
            let mut value = &config;
            if let Some(key) = &self.key {
                for k in key.split('.') {
                    value = value.get(k).ok_or_else(|| {
                        eyre::eyre!("Key not found: {} in {}", key, display_path(&file))
                    })?;
                }
            }

            match value {
                toml::Value::String(s) => miseprintln!("{}", s),
                toml::Value::Integer(i) => miseprintln!("{}", i),
                toml::Value::Boolean(b) => miseprintln!("{}", b),
                toml::Value::Float(f) => miseprintln!("{}", f),
                toml::Value::Datetime(d) => miseprintln!("{}", d),
                toml::Value::Array(a) => {
                    // seems that the toml crate does not have a way to serialize an array directly?
                    // workaround which only handle non-nested arrays
                    let elements: Vec<String> = a
                        .iter()
                        .map(|v| match v {
                            toml::Value::String(s) => format!("\"{}\"", s),
                            toml::Value::Integer(i) => i.to_string(),
                            toml::Value::Boolean(b) => b.to_string(),
                            toml::Value::Float(f) => f.to_string(),
                            toml::Value::Datetime(d) => d.to_string(),
                            toml::Value::Array(_) => "[...]".to_string(),
                            toml::Value::Table(_) => "{...}".to_string(),
                        })
                        .collect();
                    miseprintln!("[{}]", elements.join(", "));
                }
                toml::Value::Table(t) => {
                    miseprintln!("{}", toml::to_string(t)?);
                }
            }
        } else {
            bail!("No mise.toml file found");
        }
        Ok(())
    }
}

static AFTER_LONG_HELP: &str = color_print::cstr!(
    r#"<bold><underline>Examples:</underline></bold>

    $ <bold>mise toml get tools.python</bold>
    3.12
"#
);

#[cfg(test)]
mod tests {
    use crate::test::reset;

    #[test]
    fn test_toml_get() {
        reset();
        assert_cli_snapshot!("cfg", "get");
        assert_cli_snapshot!("cfg", "get", "env.TEST_ENV_VAR", @"test-123");
    }
}
