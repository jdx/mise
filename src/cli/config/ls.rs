use console::style;
use itertools::Itertools;
use miette::Result;
use tabled::settings::object::Columns;

use tabled::settings::{Modify, Width};
use tabled::Tabled;

use crate::config::config_file::ConfigFile;
use crate::config::{Config, Settings};
use crate::file::display_path;
use crate::ui::table;

/// [experimental] List config files currently in use
#[derive(Debug, clap::Args)]
#[clap(verbatim_doc_comment, after_long_help = AFTER_LONG_HELP)]
pub struct ConfigLs {
    /// Do not print table header
    #[clap(long, alias = "no-headers", verbatim_doc_comment)]
    pub no_header: bool,
}

impl ConfigLs {
    pub fn run(self) -> Result<()> {
        let config = Config::try_get()?;
        let settings = Settings::try_get()?;
        settings.ensure_experimental()?;
        let rows = config
            .config_files
            .values()
            .map(|cf| cf.as_ref().into())
            .collect::<Vec<Row>>();
        let mut table = tabled::Table::new(rows);
        table::default_style(&mut table, self.no_header);
        table.with(Modify::new(Columns::last()).with(Width::truncate(40).suffix("â€¦")));
        miseprintln!("{table}");

        Ok(())
    }
}

fn format_plugin_cell(s: String) -> String {
    match s.is_empty() {
        true => style("(none)").italic().dim().to_string(),
        false => style(s).dim().to_string(),
    }
}

#[derive(Tabled)]
#[tabled(rename_all = "PascalCase")]
struct Row {
    path: String,
    plugins: String,
}

impl From<&dyn ConfigFile> for Row {
    fn from(cf: &dyn ConfigFile) -> Self {
        let path = display_path(cf.get_path());
        let ts = cf.to_toolset();
        let plugins = ts.list_plugins().into_iter().join(", ");
        let plugins = format_plugin_cell(plugins);
        Self { path, plugins }
    }
}

// TODO: fill this out
static AFTER_LONG_HELP: &str = color_print::cstr!(
    r#"<bold><underline>Examples:</underline></bold>
  $ <bold>mise config ls</bold>
"#
);

#[cfg(test)]
mod tests {
    use std::env;

    #[test]
    fn test_config_ls() {
        assert_cli_snapshot!("cfg", "--no-headers", @r###"
        ~/cwd/.test-tool-versions tiny       
        ~/.test-tool-versions     tiny, dummy
        ~/config/config.toml      (none)
        "###);
    }
}
