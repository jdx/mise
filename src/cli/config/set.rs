use crate::cli::config::top_toml_config;
use crate::config::config_file::mise_toml::MiseToml;
use crate::config::settings::{SettingsType, SETTINGS_META};
use clap::ValueEnum;
use eyre::bail;
use std::path::PathBuf;

/// Set the value of a setting in a mise.toml file
#[derive(Debug, clap::Args)]
#[clap(after_long_help = AFTER_LONG_HELP, verbatim_doc_comment)]
pub struct ConfigSet {
    /// The path of the config to display
    pub key: String,

    /// The value to set the key to
    pub value: String,

    /// The path to the mise.toml file to edit
    ///
    /// If not provided, the nearest mise.toml file will be used
    #[clap(short, long)]
    pub file: Option<PathBuf>,

    #[clap(value_enum, short, long, default_value_t)]
    pub type_: TomlValueTypes,
}

#[derive(ValueEnum, Default, Clone, Debug)]
pub enum TomlValueTypes {
    #[default]
    Infer,
    #[value()]
    String,
    #[value()]
    Integer,
    #[value()]
    Float,
    #[value()]
    Bool,
    #[value()]
    List,
}

impl ConfigSet {
    pub fn run(self) -> eyre::Result<()> {
        let mut file = self.file;
        if file.is_none() {
            file = top_toml_config();
        }
        if let Some(file) = file {
            let mut config: toml_edit::DocumentMut = std::fs::read_to_string(&file)?.parse()?;
            let mut container = config.as_item_mut();
            let parts = self.key.split('.').collect::<Vec<&str>>();
            for key in parts.iter().take(parts.len() - 1) {
                container = container
                    .as_table_mut()
                    .unwrap()
                    .entry(key)
                    .or_insert(toml_edit::Item::Table(toml_edit::Table::new()));
            }
            let last_key = parts.last().unwrap();

            let type_to_use = match self.type_ {
                TomlValueTypes::Infer => {
                    let expected_type = if !self.key.starts_with("settings.") {
                        None
                    } else {
                        SETTINGS_META.get(&(*last_key).to_string())
                    };
                    match expected_type {
                        Some(meta) => match meta.type_ {
                            SettingsType::Bool => TomlValueTypes::Bool,
                            SettingsType::String => TomlValueTypes::String,
                            SettingsType::Integer => TomlValueTypes::Integer,
                            SettingsType::Duration => TomlValueTypes::String,
                            SettingsType::Path => TomlValueTypes::String,
                            SettingsType::Url => TomlValueTypes::String,
                            SettingsType::ListString => TomlValueTypes::List,
                            SettingsType::ListPath => TomlValueTypes::List,
                        },
                        None => TomlValueTypes::String,
                    }
                }
                _ => self.type_,
            };

            let value = match type_to_use {
                TomlValueTypes::String => toml_edit::value(self.value),
                TomlValueTypes::Integer => toml_edit::value(self.value.parse::<i64>()?),
                TomlValueTypes::Float => toml_edit::value(self.value.parse::<f64>()?),
                TomlValueTypes::Bool => toml_edit::value(self.value.parse::<bool>()?),
                TomlValueTypes::List => {
                    let mut list = toml_edit::Array::new();
                    for item in self.value.split(',').map(|s| s.trim()) {
                        list.push(item);
                    }
                    toml_edit::Item::Value(toml_edit::Value::Array(list))
                }
                TomlValueTypes::Infer => bail!("Type not found"),
            };

            container.as_table_mut().unwrap().insert(last_key, value);

            let raw = config.to_string();
            MiseToml::from_str(&raw, &file)?;
            std::fs::write(&file, raw)?;
        } else {
            bail!("No mise.toml file found");
        }
        Ok(())
    }
}

static AFTER_LONG_HELP: &str = color_print::cstr!(
    r#"<bold><underline>Examples:</underline></bold>

    $ <bold>mise config set tools.python 3.12</bold>
    $ <bold>mise config set settings.always_keep_download true</bold>
    $ <bold>mise config set env.TEST_ENV_VAR ABC</bold>
    $ <bold>mise config set settings.disable_tools --type list node,rust</bold>

    # Type for `settings` is inferred
    $ <bold>mise config set settings.jobs 4</bold>
"#
);

#[cfg(test)]
mod tests {
    use crate::test::reset;

    #[test]
    fn test_config_set() {
        reset();
        assert_cli_snapshot!("config", "set", "env.TEST_ENV_VAR", "ABC", @"");
        assert_cli_snapshot!("config", "get", "env.TEST_ENV_VAR", @"ABC");

        assert_cli_snapshot!("config", "set", "settings.ruby.default_packages_file", "abc", @"");
        assert_cli_snapshot!("config", "get", "settings.ruby.default_packages_file", @"abc");

        assert_cli_snapshot!("config", "set", "settings.always_keep_download", "--type", "bool", "true", @"");
        assert_cli_snapshot!("config", "get", "settings.always_keep_download", @"true");

        assert_cli_snapshot!("config", "set", "settings.jobs", "--type", "integer", "4", @"");
        assert_cli_snapshot!("config", "get", "settings.jobs", @"4");

        assert_cli_snapshot!("config", "set", "settings.jobs", "4", @"");
        assert_cli_snapshot!("config", "get", "settings.jobs", @"4");

        assert_cli_snapshot!("config", "set", "settings.disable_tools", "--type", "list", "node,rust", @"");
        assert_cli_snapshot!("config", "get", "settings.disable_tools", @"[\"node\", \"rust\"]");
    }
}
