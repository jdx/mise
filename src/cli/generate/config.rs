use crate::Result;
use crate::cli::config::generate;

/// [experimental] Generate a mise.toml file
#[derive(Debug, clap::Args)]
#[clap(verbatim_doc_comment, after_long_help = generate::AFTER_LONG_HELP)]
pub struct Config {
    #[clap(flatten)]
    generate: generate::ConfigGenerate,
}

impl Config {
    pub async fn run(self) -> Result<()> {
        self.generate.run()
    }
}
