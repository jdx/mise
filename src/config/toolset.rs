use std::collections::HashMap;
use std::sync::Arc;

use color_eyre::eyre::Result;
use indexmap::IndexMap;
use itertools::Itertools;
use rayon::prelude::*;
use versions::Mess;

use crate::config::AliasMap;
use crate::plugins::{Plugin, PluginName, PluginSource};
use crate::runtimes::RuntimeVersion;

#[derive(Debug, Default)]
pub struct Toolset {
    pub plugins: HashMap<PluginName, Arc<Plugin>>,
    installed_versions: HashMap<PluginName, HashMap<String, Arc<RuntimeVersion>>>,
    current_versions: IndexMap<PluginName, Vec<String>>,
    current_versions_sources: HashMap<PluginName, PluginSource>,
}

impl Toolset {
    pub fn find_plugin(&self, key: &PluginName) -> Option<Arc<Plugin>> {
        self.plugins.get(key).map(Arc::clone)
    }

    pub fn get_or_add_plugin(&mut self, name: String) -> Result<Arc<Plugin>> {
        let plugin = match self.plugins.get(&name) {
            Some(p) => p,
            None => {
                let plugin = Plugin::load(&name)?;
                self.plugins.entry(name).or_insert_with(|| Arc::new(plugin))
            }
        };

        Ok(plugin.clone())
    }

    pub fn get_or_add_version(
        &mut self,
        plugin: &str,
        version: String,
    ) -> Result<Arc<RuntimeVersion>> {
        let plugin = self.get_or_add_plugin(plugin.into())?;
        let rtv = self
            .installed_versions
            .entry(plugin.name.clone())
            .or_default()
            .entry(version.clone())
            .or_insert_with(|| Arc::new(RuntimeVersion::new(plugin, &version)))
            .clone();

        Ok(rtv)
    }

    pub fn add_runtime_versions(&mut self, plugin: &str, versions: Vec<String>) -> Result<()> {
        for version in versions {
            self.get_or_add_version(plugin, version)?;
        }
        Ok(())
    }

    pub fn set_current_runtime_versions(
        &mut self,
        plugin: &str,
        versions: Vec<String>,
        source: PluginSource,
    ) -> Result<()> {
        self.get_or_add_plugin(plugin.into())?;
        self.current_versions.insert(plugin.into(), versions);
        self.current_versions_sources.insert(plugin.into(), source);
        Ok(())
    }

    pub fn list_plugins(&self) -> Vec<Arc<Plugin>> {
        self.plugins
            .values()
            .sorted_by_cached_key(|p| p.name.clone())
            .map(Arc::clone)
            .collect()
    }

    pub fn list_installed_plugins(&self) -> Vec<Arc<Plugin>> {
        self.list_plugins()
            .into_iter()
            .filter(|p| p.is_installed())
            .collect()
    }

    pub fn list_current_plugins(&self) -> Vec<Arc<Plugin>> {
        self.current_versions
            .keys()
            .sorted()
            .map(|p| self.find_plugin(p).unwrap())
            .collect()
    }

    pub fn list_installed_versions(&self) -> Vec<Arc<RuntimeVersion>> {
        self.installed_versions
            .iter()
            .sorted_by_cached_key(|(plugin_name, _)| plugin_name.to_string())
            .flat_map(|(_, versions)| versions.iter().map(|(_, rtv)| rtv.clone()))
            .collect()
    }

    pub fn list_current_versions(&self) -> Vec<Arc<RuntimeVersion>> {
        self.current_versions
            .iter()
            .flat_map(|(plugin_name, versions)| {
                versions
                    .iter()
                    .map(|v| {
                        self.resolve_version(plugin_name, v).unwrap_or_else(|| {
                            let plugin = self
                                .find_plugin(plugin_name)
                                .unwrap_or_else(|| Arc::new(Plugin::new(plugin_name)));
                            Arc::new(RuntimeVersion::new(plugin, v))
                        })
                    })
                    .collect::<Vec<_>>()
            })
            .collect()
    }

    pub fn list_current_installed_versions(&self) -> Vec<Arc<RuntimeVersion>> {
        self.list_current_versions()
            .into_iter()
            .filter(|rtv| rtv.is_installed())
            .collect()
    }

    pub fn resolve_all_versions(&mut self, aliases: &AliasMap) -> Result<()> {
        let default_aliases = IndexMap::new();
        self.current_versions = self
            .current_versions
            .clone()
            .into_iter()
            .collect_vec()
            .into_par_iter()
            .map(|(plugin_name, versions)| {
                let aliases = aliases.get(&plugin_name).unwrap_or(&default_aliases);
                let plugin = self
                    .find_plugin(&plugin_name)
                    .unwrap_or_else(|| Arc::new(Plugin::new(&plugin_name)));
                let versions = versions
                    .iter()
                    .map(|v| {
                        let v = match aliases.get(v) {
                            Some(version) => {
                                trace!("resolved alias: {}@{} -> {}", plugin.name, v, version);
                                version
                            }
                            _ => v,
                        };
                        match self.resolve_version(&plugin_name, v) {
                            Some(rtv) => Ok(rtv.version.clone()),
                            None => {
                                let latest = if plugin.is_installed() {
                                    plugin.latest_version(v)?
                                } else {
                                    Some(v.clone())
                                };
                                Ok(latest.unwrap_or_else(|| v.clone()))
                            }
                        }
                    })
                    .collect::<Result<Vec<String>>>()?;
                Ok((plugin_name, versions))
            })
            .collect::<Result<Vec<(PluginName, Vec<String>)>>>()?
            .into_iter()
            .collect::<IndexMap<PluginName, Vec<String>>>();
        trace!("resolved versions: {:?}", self.current_versions);
        Ok(())
        // if plugin.is_installed() {
        //     if let Some(latest) = plugin.latest_version(version)? {
        //         return Ok(Arc::new(RuntimeVersion::new(plugin, &latest)));
        //     }
        // }
        // Ok(Arc::new(RuntimeVersion::new(plugin, version)))
    }

    pub fn resolve_version(
        &self,
        plugin: &PluginName,
        version: &str,
    ) -> Option<Arc<RuntimeVersion>> {
        if let Some(installed_versions) = self.installed_versions.get(plugin) {
            if let Some(rtv) = installed_versions.get(version) {
                return Some(rtv.clone());
            }
            let sorted_versions = installed_versions
                .keys()
                .sorted_by_cached_key(|v| v.parse::<Mess>().unwrap())
                .rev()
                .collect::<Vec<_>>();
            for v in sorted_versions {
                if v.starts_with(version) {
                    return Some(installed_versions[v].clone());
                }
            }
        }

        None
    }

    pub fn get_source_for_plugin(&self, plugin: &PluginName) -> Option<PluginSource> {
        self.current_versions_sources.get(plugin).cloned()
    }

    pub fn find_by_prefix(
        &self,
        aliases: &AliasMap,
        plugin: &str,
        prefix: &str,
    ) -> Option<Arc<RuntimeVersion>> {
        let default_aliases = IndexMap::new();
        let aliases = aliases.get(plugin).unwrap_or(&default_aliases);
        let prefix = aliases.get(prefix).cloned().unwrap_or(prefix.to_string());

        let mut versions = self.list_current_versions();
        versions.extend(self.list_installed_versions());
        versions
            .into_iter()
            .find(|rtv| rtv.plugin.name == plugin && rtv.version.starts_with(&prefix))
    }
}
