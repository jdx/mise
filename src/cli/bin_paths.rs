use color_eyre::eyre::Result;

use crate::config::Config;

use crate::toolset::ToolsetBuilder;

/// List all the active runtime bin paths
#[derive(Debug, clap::Args)]
#[clap(verbatim_doc_comment)]
pub struct BinPaths {}

impl BinPaths {
    pub fn run(self, config: &Config) -> Result<()> {
        let ts = ToolsetBuilder::new().build(config)?;
        for p in ts.list_paths() {
            rtxprintln!("{}", p.display());
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use crate::{assert_cli, assert_cli_snapshot};

    #[test]
    fn test_bin_paths() {
        assert_cli!("i");
        assert_cli_snapshot!("bin-paths");
    }
}
