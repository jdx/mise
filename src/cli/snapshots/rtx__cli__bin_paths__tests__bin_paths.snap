---
source: src/cli/bin_paths.rs
expression: output
---
~/data/installs/tiny/3.1.0/bin

